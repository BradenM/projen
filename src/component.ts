import { Construct } from "constructs";

/**
 * Represents a project component.
 */
export class Component extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  /**
   * Called before synthesis.
   */
  public preSynthesize() {}

  /**
   * Synthesizes files to the project output directory.
   */
  public synthesize() {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   */
  public postSynthesize() {}
}
