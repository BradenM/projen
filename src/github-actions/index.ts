export * from "./actions";
export * from "./github-action-metadata";
