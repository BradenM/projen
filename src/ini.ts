import { Construct } from "constructs";
import * as INI from "ini";
import { IResolver } from "./file";
import { ObjectFile, ObjectFileOptions } from "./object-file";

/**
 * Options for `IniFile`.
 */
export interface IniFileOptions extends ObjectFileOptions {}

/**
 * Represents an INI file.
 */
export class IniFile extends ObjectFile {
  constructor(scope: Construct, filePath: string, options: IniFileOptions) {
    super(scope, filePath, options);
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const json = super.synthesizeContent(resolver);
    if (!json) {
      return undefined;
    }

    return [
      ...(this.marker ? [`# ${this.marker}`] : []),
      "",
      INI.stringify(JSON.parse(json)),
    ].join("\n");
  }
}
