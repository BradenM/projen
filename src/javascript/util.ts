import { readFileSync } from "fs";
import { basename, dirname, extname, join, sep } from "path";
import * as semver from "semver";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>\w+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifestId = `${dependencyName}/package.json`;
  try {
    const manifestPath = require.resolve(manifestId, options);
    // cannot just `require('dependency/package.json')` here because
    // `options.paths` may not overlap with this node proc's resolution paths.
    const manifest = JSON.parse(
      readFileSync(manifestPath, "utf-8").toString()
    ) as {
      version: string;
    };
    return manifest.version;
  } catch (e) {
    return undefined;
  }
}
