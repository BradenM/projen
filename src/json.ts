import { Construct } from "constructs";
import { IResolver } from "./file";
import { ObjectFile, ObjectFileOptions } from "./object-file";

/**
 * Options for `JsonFile`.
 */
export interface JsonFileOptions extends ObjectFileOptions {
  /**
   * Adds a newline at the end of the file.
   * @default true
   */
  readonly newline?: boolean;
}

/**
 * Represents a JSON file.
 */
export class JsonFile extends ObjectFile {
  private readonly newline: boolean;

  constructor(scope: Construct, filePath: string, options: JsonFileOptions) {
    super(scope, filePath, options);

    this.newline = options.newline ?? true;

    if (!options.obj) {
      throw new Error('"obj" cannot be undefined');
    }
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const json = super.synthesizeContent(resolver);
    if (!json) {
      return undefined;
    }

    const sanitized = JSON.parse(json);

    if (this.marker) {
      sanitized["//"] = this.marker;
    }

    let content = JSON.stringify(sanitized, undefined, 2);
    if (this.newline) {
      content += "\n";
    }

    return content;
  }
}
