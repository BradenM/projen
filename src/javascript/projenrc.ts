import { existsSync, writeFileSync } from "fs";
import { resolve } from "path";
import { Construct } from "constructs";
import { Component } from "../component";
import { Project } from "../project";
import { renderJavaScriptOptions } from "./render-options";
export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.js"
   */
  readonly filename?: string;
}

/**
 * Sets up a javascript project to use TypeScript for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(scope: Construct, options: ProjenrcOptions = {}) {
    super(scope, "Projenrc");

    const project = Project.of(this);

    this.rcfile = options.filename ?? ".projenrc.js";

    // this is the task projen executes when running `projen`
    project.defaultTask?.exec(`node ${this.rcfile}`);

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(Project.of(this).outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = Project.of(this).initProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split(".");
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join(".");

    const { renderedOptions, imports } = renderJavaScriptOptions({
      comments: bootstrap.comments,
      args: bootstrap.args,
      type: bootstrap.type,
    });

    imports.add(importName);

    const lines = new Array<string>();
    lines.push(
      `const { ${[...imports].sort().join(", ")} } = require("${moduleName}");`
    );
    lines.push();
    lines.push(`const project = new ${className}(${renderedOptions});`);
    lines.push();
    lines.push("project.synth();");

    writeFileSync(rcfile, lines.join("\n"));
    Project.of(this).logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
