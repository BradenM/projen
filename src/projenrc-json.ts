import { existsSync, writeFileSync } from "fs";
import { resolve } from "path";
import { Construct } from "constructs";
import { Component } from "./component";
import { Project } from "./project";

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.json"
   */
  readonly filename?: string;
}

/**
 * Sets up a project to use JSON for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(scope: Construct, options: ProjenrcOptions = {}) {
    super(scope, "Projenrc");

    this.rcfile = options.filename ?? ".projenrc.json";

    // this is the task projen executes when running `projen`
    Project.of(this).defaultTask?.env("FILENAME", this.rcfile);
    Project.of(this).defaultTask?.builtin("run-projenrc-json");

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(Project.of(this).outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = Project.of(this).initProject;
    if (!bootstrap) {
      return;
    }

    const json = {
      type: bootstrap.fqn,
      ...bootstrap.args,
    };

    writeFileSync(rcfile, JSON.stringify(json, null, 2));
    Project.of(this).logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
