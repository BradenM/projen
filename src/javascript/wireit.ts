import * as path from "path";
import { Component } from "../component";
import { Project } from "../project";
import { NodePackageManager } from "./node-package";
import { NodeProject } from "./node-project";
import { determineLockfile } from "./util";

export interface WireitOptions {
  /**
   * The Node Package Manager used to execute scripts
   *
   * @default NodePackageManager.YARN
   */
  readonly packageManager?: NodePackageManager;

  readonly projenCommand?: string;
}

export class Wireit extends Component {
  public static of(project: Project): Wireit | undefined {
    const isWireit = (c: Component): c is Wireit => c instanceof Wireit;
    return project.components.find(isWireit);
  }

  private readonly nodeProject: NodeProject;
  private readonly packageManager: NodePackageManager;

  constructor(project: NodeProject, options: WireitOptions) {
    super(project);

    this.nodeProject = project;
    this.packageManager = options.packageManager ?? NodePackageManager.YARN;

    project.addGitIgnore(".wireit");
  }

  public renderConfig(): any {
    const tasks: { [name: string]: WireitTaskSpec } = {};
    for (const task of this.project.tasks.all) {
      const steps = task.steps.map((step) => {
        let command = "";
        if (step.cwd) {
          command += `cd ${step.cwd} && `;
        }
        if (step.say) {
          command += `echo ${step.say}`;
        }
        if (step.builtin) {
          command += renderBuiltin(step.builtin);
        }
        if (step.exec) {
          command += step.exec;
        }
        if (step.spawn) {
          const subtask = this.nodeProject.tasks.tryFind(step.spawn);
          if (!subtask) {
            throw new Error(
              `Could not find task ${subtask} (spawned by ${task.name})`
            );
          }
          command += this.nodeProject.runTaskCommand(subtask);
        }
        return command;
      });
      tasks[task.name] = {
        command:
          steps.length > 0
            ? steps.join(" && ")
            : // If there is no command and no dependencies, wireit will error, so we add a placeholder command.
            task.dependencies.length === 0
            ? "echo No commands to run."
            : undefined,
        clean: task.clean,
        files: task.inputs,
        output: task.outputs,
        dependencies: task.dependencies,
        packageLocks: [determineLockfile(this.packageManager)],
      };
    }
    return tasks;
  }
}

export interface WireitTaskSpec {
  readonly command?: string;
  readonly dependencies?: string[];
  readonly files?: string[];
  readonly output?: string[];
  readonly clean?: boolean;
  readonly packageLocks?: string[];
}

function renderBuiltin(builtin: string) {
  const moduleRoot = path.dirname(require.resolve("../../package.json"));
  const program = require.resolve(
    path.join(moduleRoot, "lib", `${builtin}.task.js`)
  );
  return `${process.execPath} ${program}`;
}
