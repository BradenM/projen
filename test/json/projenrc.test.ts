import { StandardProject, Testing } from "../../src";
import { renderProjenInitOptions } from "../../src/javascript/render-options";
import { Projenrc } from "../../src/projenrc-json";

test("projenrc.json default project", () => {
  // GIVEN
  const project = new StandardProject({ name: "my-project" });

  // WHEN
  new Projenrc(project);

  // THEN
  expect(Testing.synth(project)).toMatchSnapshot();
});

test("projenrc.json with typed options", () => {
  // GIVEN
  const project = new StandardProject(
    renderProjenInitOptions("projen.typescript.TypeScriptProject", {
      staleOptions: {
        issues: {
          daysBeforeStale: 100, // number, nested option
        },
        pullRequest: {
          enabled: false, // boolean, nested option
        },
      },
      name: "@example/foo", // string
      npmAccess: "public", // enum provided as string
    })
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(Testing.synth(project)).toMatchSnapshot();
});
