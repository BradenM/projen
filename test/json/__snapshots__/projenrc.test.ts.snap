// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projenrc.json default project 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/stale.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v3.4.6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: 14
          days-before-pr-close: 2
          stale-pr-message: This pull request is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-pr-message: Closing this pull request as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-pr-label: stale
          exempt-pr-labels: backlog
          days-before-issue-stale: 60
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-issue-message: Closing this issue as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-issue-label: stale
          exempt-issue-labels: backlog
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/stale.yml
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/stale.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "env": Object {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": Array [
          Object {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  ".projen/tree.json": Object {
    "tree": Object {
      "children": Object {
        ".gitattributes": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".gitattributes",
          "path": "my-project/.gitattributes",
        },
        ".gitignore": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".gitignore",
          "path": "my-project/.gitignore",
        },
        ".projen--files.json": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".projen--files.json",
          "path": "my-project/.projen--files.json",
        },
        "Dependencies": Object {
          "children": Object {
            ".projen--deps.json": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": ".projen--deps.json",
              "path": "my-project/Dependencies/.projen--deps.json",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Dependencies",
          "path": "my-project/Dependencies",
        },
        "GitHub": Object {
          "children": Object {
            "Mergify": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "Mergify",
              "path": "my-project/GitHub/Mergify",
            },
            "PullRequestLint": Object {
              "children": Object {
                "pull-request-lint": Object {
                  "children": Object {
                    ".github--workflows--pull-request-lint.yml": Object {
                      "constructInfo": Object {
                        "fqn": "constructs.Construct",
                        "version": "10.0.78",
                      },
                      "id": ".github--workflows--pull-request-lint.yml",
                      "path": "my-project/GitHub/PullRequestLint/pull-request-lint/.github--workflows--pull-request-lint.yml",
                    },
                  },
                  "constructInfo": Object {
                    "fqn": "constructs.Construct",
                    "version": "10.0.78",
                  },
                  "id": "pull-request-lint",
                  "path": "my-project/GitHub/PullRequestLint/pull-request-lint",
                },
              },
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "PullRequestLint",
              "path": "my-project/GitHub/PullRequestLint",
            },
            "Stale": Object {
              "children": Object {
                "stale": Object {
                  "children": Object {
                    ".github--workflows--stale.yml": Object {
                      "constructInfo": Object {
                        "fqn": "constructs.Construct",
                        "version": "10.0.78",
                      },
                      "id": ".github--workflows--stale.yml",
                      "path": "my-project/GitHub/Stale/stale/.github--workflows--stale.yml",
                    },
                  },
                  "constructInfo": Object {
                    "fqn": "constructs.Construct",
                    "version": "10.0.78",
                  },
                  "id": "stale",
                  "path": "my-project/GitHub/Stale/stale",
                },
              },
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "Stale",
              "path": "my-project/GitHub/Stale",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "GitHub",
          "path": "my-project/GitHub",
        },
        "Logger": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Logger",
          "path": "my-project/Logger",
        },
        "ProjectBuild": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "ProjectBuild",
          "path": "my-project/ProjectBuild",
        },
        "Projenrc": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Projenrc",
          "path": "my-project/Projenrc",
        },
        "README.md": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "README.md",
          "path": "my-project/README.md",
        },
        "Tasks": Object {
          "children": Object {
            ".projen--tasks.json": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": ".projen--tasks.json",
              "path": "my-project/Tasks/.projen--tasks.json",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Tasks",
          "path": "my-project/Tasks",
        },
        "VsCode": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "VsCode",
          "path": "my-project/VsCode",
        },
      },
      "constructInfo": Object {
        "fqn": "constructs.Construct",
        "version": "10.0.78",
      },
      "id": "my-project",
      "path": "my-project",
    },
    "version": "tree-0.1",
  },
  "README.md": "# replace this",
}
`;

exports[`projenrc.json with typed options 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/stale.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v3.4.6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: -1
          days-before-pr-close: -1
          days-before-issue-stale: 100
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-issue-message: Closing this issue as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-issue-label: stale
          exempt-issue-labels: backlog
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/stale.yml
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/stale.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "env": Object {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": Array [
          Object {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  ".projen/tree.json": Object {
    "tree": Object {
      "children": Object {
        ".gitattributes": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".gitattributes",
          "path": "@example--foo/.gitattributes",
        },
        ".gitignore": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".gitignore",
          "path": "@example--foo/.gitignore",
        },
        ".projen--files.json": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": ".projen--files.json",
          "path": "@example--foo/.projen--files.json",
        },
        "Dependencies": Object {
          "children": Object {
            ".projen--deps.json": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": ".projen--deps.json",
              "path": "@example--foo/Dependencies/.projen--deps.json",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Dependencies",
          "path": "@example--foo/Dependencies",
        },
        "GitHub": Object {
          "children": Object {
            "Mergify": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "Mergify",
              "path": "@example--foo/GitHub/Mergify",
            },
            "PullRequestLint": Object {
              "children": Object {
                "pull-request-lint": Object {
                  "children": Object {
                    ".github--workflows--pull-request-lint.yml": Object {
                      "constructInfo": Object {
                        "fqn": "constructs.Construct",
                        "version": "10.0.78",
                      },
                      "id": ".github--workflows--pull-request-lint.yml",
                      "path": "@example--foo/GitHub/PullRequestLint/pull-request-lint/.github--workflows--pull-request-lint.yml",
                    },
                  },
                  "constructInfo": Object {
                    "fqn": "constructs.Construct",
                    "version": "10.0.78",
                  },
                  "id": "pull-request-lint",
                  "path": "@example--foo/GitHub/PullRequestLint/pull-request-lint",
                },
              },
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "PullRequestLint",
              "path": "@example--foo/GitHub/PullRequestLint",
            },
            "Stale": Object {
              "children": Object {
                "stale": Object {
                  "children": Object {
                    ".github--workflows--stale.yml": Object {
                      "constructInfo": Object {
                        "fqn": "constructs.Construct",
                        "version": "10.0.78",
                      },
                      "id": ".github--workflows--stale.yml",
                      "path": "@example--foo/GitHub/Stale/stale/.github--workflows--stale.yml",
                    },
                  },
                  "constructInfo": Object {
                    "fqn": "constructs.Construct",
                    "version": "10.0.78",
                  },
                  "id": "stale",
                  "path": "@example--foo/GitHub/Stale/stale",
                },
              },
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": "Stale",
              "path": "@example--foo/GitHub/Stale",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "GitHub",
          "path": "@example--foo/GitHub",
        },
        "Logger": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Logger",
          "path": "@example--foo/Logger",
        },
        "ProjectBuild": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "ProjectBuild",
          "path": "@example--foo/ProjectBuild",
        },
        "Projenrc": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Projenrc",
          "path": "@example--foo/Projenrc",
        },
        "README.md": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "README.md",
          "path": "@example--foo/README.md",
        },
        "Tasks": Object {
          "children": Object {
            ".projen--tasks.json": Object {
              "constructInfo": Object {
                "fqn": "constructs.Construct",
                "version": "10.0.78",
              },
              "id": ".projen--tasks.json",
              "path": "@example--foo/Tasks/.projen--tasks.json",
            },
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "Tasks",
          "path": "@example--foo/Tasks",
        },
        "VsCode": Object {
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.78",
          },
          "id": "VsCode",
          "path": "@example--foo/VsCode",
        },
      },
      "constructInfo": Object {
        "fqn": "constructs.Construct",
        "version": "10.0.78",
      },
      "id": "@example--foo",
      "path": "@example--foo",
    },
    "version": "tree-0.1",
  },
  ".projenrc.json": Object {
    "name": "@example/foo",
    "npmAccess": "public",
    "staleOptions": Object {
      "issues": Object {
        "daysBeforeStale": 100,
      },
      "pullRequest": Object {
        "enabled": false,
      },
    },
    "type": "projen.typescript.TypeScriptProject",
  },
  "README.md": "# replace this",
}
`;
